RecoParticleFlow/PFClusterProducer/plugins/LocalMaximumSeedFinder.cc

```diff
@@ -93,7 +95,8 @@ LocalMaximumSeedFinder::LocalMaximumSeedFinder(const edm::ParameterSet& conf)
 // the starting state of seedable is all false!
 void LocalMaximumSeedFinder::findSeeds(const edm::Handle<reco::PFRecHitCollection>& input,
                                        const std::vector<bool>& mask,
-                                       std::vector<bool>& seedable) {
+                                       std::vector<bool>& seedable,
+                                       HcalPFCuts* hcalCuts) {
```

Comment:
Naively, rather than percolating this as a pointer argument through many functions, I would have expected some-kind-of common utility class (e.g. "PFThresholdHelper") which (a) is a class member of each client (one client being `LocalMaximumSeedFinder`, for example), (b) gets initialised in the constructor of the client (if not in the constructor, in the method which accesses the `edm::ParameterSet`), and (c) can return the thresholds wherever the client needs them. When the utility class is built, it could figure out, based on the value of `PFcutsFromDB` from the PSet, if it should use the EventSetup token, or look for the thresholds in the PSet.

This leads me to another question. If `PFcutsFromDB = True`, does one still have to put "some" thresholds in the PSet (the python configuration) the same way we do now ? I'm missing the part where `PFcutsFromDB = True` is used to avoid looking for the thresholds in the PSet. Maybe it's just me, but I'm worried about having to keep a bunch of unused (or dummy) numbers in the PSets of (the modules of) the HLT menus, even after HLT moves to `PFcutsFromDB = True`.

Reply:
@swagata87 please clarify if these comments are going to be followed up

Reply:
Thanks Marino. 
The same code is used for ECAL also, and ECAL still pass the threshold values from config files. Also Phase2 HCAL still uses thresholds from config file. So the threshold parameters can't be removed right now from PSet. 

As we are in a bit of rush (we want to make sure this GT based approach is usable in 2024 data-taking, and it gets into a release soon so that some validations can be performed by PdmV), so I implemented it in a way that is minimally invasive, and also easy/straightforward for me to implement it. Once we get to a point where it is in a working condition and producing exactly same physics results as before, we can surely think about code-related improvements. Actually many places in PF code needs improvement, and we are trying to do it in parallel (depends a bit on available time and personpower). So if you agree, if this approach is working, lets move forward with this. We can come back to coding style etc later on once the immediate rush is over.

Reply:
> Once we get to a point where it is in a working condition and producing exactly same physics results as before

I am not sure to follow. This changes the behaviour by default, so wasn't the change proposed here already tested to not alter the physics?

Reply:
sorry I wasn't clear. Indeed the codes that I migrated in this PR seem to be working fine. From PR test we can see that there is no change in physics (later PdmV can validate on more stat of course). But there are other codes still to be migrated to use the GT based approach. From PF side, we will migrate the PF recHit producer codes. And some POG codes also need to be adapted (I think mostly EGM POG). So we can focus on this migration for now, and do it asap. Once all migration are done then we can come back to the improvement aspects.

Reply:
The migration of other PF codes can indeed be done as part of this PR. I was thinking of a separate PR as this one already has a long change-log. But your point is also important, if we have to trace back any issue that might show up later, then having all changes in 1 PR is better. 

The EGM part (or other POGs if any) need to be a separate, as other people will implement those.

Reply:
> (and of course if @missirol agrees with that)

Sure, I have no objection.

Reply:
> And some POG codes also need to be adapted (I think mostly EGM POG).

@cms-sw/hcal-dpg-l2 

I only have one more question on this (although this might not be the right place to ask it, but I want to check if I'm following). HLT also uses (in various ways) the plugin `CaloTowersCreator`. These are the so-called "calo towers". HCal has asked to move away from this a while ago (in favour of PFRecHit, iiuc), but this has not happened yet for many POGs (JME, TAU, maybe others). Afaik, `CaloTowersCreator` also uses the thresholds discussed here via its PSet, e.g. [*].

Is this correct ? Will / could / should `CaloTowersCreator` be updated too ? (in a separate PR, of course)

Ultimately, I don't think we would want a situation where some modules of the HLT menu get the thresholds from the GT, while some other modules rely on their PSet.

[*]
```
[CaloTowersCreator].HBThreshold1
[CaloTowersCreator].HBThreshold2
[CaloTowersCreator].HBThreshold
```

Reply:
> so what do people prefer? we extend this PR to migrate other PF codes? Or we do it separately? I am okay with any of it.

I would personally prefer to do all the PF change in this PR.

Reply:
okay I will start to work on those soon

Reply:
Hi @missirol , @swagata87 and @mmusich ,
We HCAL discussed the thresholds for CaloTowers in the DPG meeting.
We can update them, but we prefer to let the downstream consumers take care of them.
For example, [here](https://cmssdt.cern.ch/lxr/source/HLTrigger/Configuration/python/HLT_2023v12_cff.py#7748) are the thresholds in HLT for 2023 data-taking:
`hltTowerMakerForAll.HBThreshold1 = cms.double( 0.4 )`
`hltTowerMakerForAll.HBThreshold2 = cms.double( 0.3 )`
`hltTowerMakerForAll.HBThreshold = cms.double( 0.3 ) `

Reply:
> We can update them, but we prefer to let the downstream consumers take care of them.

Sorry,  can you clarify what you mean by "downstream" consumers in this context? The owners of the paths that employ calotowers? That's gonna be very inefficient IMO.

Reply:
> We HCAL discussed the thresholds for CaloTowers in the DPG meeting. We can update them, but we prefer to let the downstream consumers take care of them.

On this, I didn't understand if HCAL could take care of updating `CaloTowersCreator` to make use of `HcalPFCuts`.. ?

Reply:
@wang-hui see please above.

Reply:
Sorry I don't understand your comment. CaloTower is a collection of recHits and it does not use PFcuts. Let me start an email thread for further discussions